updatecli/updatecli
# Feat: support partial content for the resource "file"

Fix #231 , depends on #321

This PR allows manipulating partial content of files.

- Source of kind `file` can now:
  - Read the whole content from a file or an URL (same as before)
  - Read only the content of the line (by specifying its number) from a file or an URL (breaking: removes the old behavior that was transfomring lines)
  - Read the content of matched string(s) (new behavior) when provided a regex pattern

- Conditions of kind `file` can now check that a file (or an URL) :
  - has its content matching the input source, or the specified content (same as before)
  - has the specified line matching the input source, or the specified content (new)
  - has the specified line not empty (new) 
  - has its content, OR the specified line, matching the provided regex pattern (new)

- Targets of kind `file` can now update a file:
  - totally, .e.g its whole content from either the source input value or specified content (same as before)
  - partially for a line, e.g. only the specified line is changed to either the source input value or specified content (new)
  - partially for a match + replace regex pattern e.g. a search and replace is applied (new)

### How to Test

- The file `examples/updateCli.generic/file.yaml` had been updated to be a "reference" of different cases
- Unit tests had been added to the plugin `file` to ensure that a lot of cases can be run in less than 1 seconds (not mentioning the ability to debug easily)

```shell
cd ./pkg/plugins/file
go test
```

- A real life test has been done with the file https://github.com/jenkins-infra/packer-images/blob/a3a5e4a7011c571a22ef6e274eefab334f70cc31/updatecli/updatecli.d/maven.yml that was changed to:

<details><summary>Click to show the code snippet</summary>

```yaml
---
title: "Bump Maven version"
sources:
  getLatestMavenVersion:
    kind: githubRelease
    name: Get the latest Maven version
    spec:
      owner: "apache"
      repository: "maven"
      token: "{{ requiredEnv .github.token }}"
      username: "{{ .github.username }}"
      versionFilter:
        kind: latest
    transformers:
      - trimPrefix: "maven-"

conditions:
  checkIfReleaseIsAvailable:
    kind: shell
    disablesourceinput: true
    spec:
      command: curl --connect-timeout 5 --location --head --fail --silent --show-error https://archive.apache.org/dist/maven/maven-3/{{ source `getLatestMavenVersion` }}/binaries/apache-maven-{{ source `getLatestMavenVersion` }}-bin.tar.gz
  checkIfMavenVersionIsDefined:
    kind: file
    disablesourceinput: true
    spec:
      file: ".auto.pkrvars.hcl"
      matchPattern: 'maven_(.*)=.*'

targets:
  updateMavenVersion:
    name: Update the Maven version in the Packer default values
    sourceID: getLatestMavenVersion
    kind: file
    spec:
      file: ".auto.pkrvars.hcl"
      matchPattern: 'maven_(.*)=.*'
      replacePattern: 'maven_$1= "{{ source `getLatestMavenVersion` }}"'
    scm:
      github:
        user: "{{ .github.user }}"
        email: "{{ .github.email }}"
        owner: "{{ .github.owner }}"
        repository: "{{ .github.repository }}"
        token: "{{ requiredEnv .github.token }}"
        username: "{{ .github.username }}"
        branch: "{{ .github.branch }}"
```
</details>

## Additional Details

- Related to the added Unit Test, the packages `shell`, `scm` and `text` are now providing a reusable Mock for some of their internals. This code is always stored within a file named `mocks.go`
